packages/data-context/src/data/ProjectConfigManager.ts

```diff
@@ -0,0 +1,490 @@
+  private _cachedLoadConfig: LoadConfigReply | undefined
+  private _cypressEnv: CypressEnv
+
+  constructor (private options: ProjectConfigManagerOptions) {
+    this._cypressEnv = new CypressEnv({
+      envFilePath: this.envFilePath,
+      validateConfigFile: (filePath, config) => {
+        this.validateConfigFile(filePath, config)
+      },
+    })
+
+    return autoBindDebug(this)
+  }
+
+  get isLoadingNodeEvents () {
+    return this._state === 'loadingNodeEvents'
+  }
+
+  get isReady () {
+    return this._state === 'ready'
+  }
+
+  get isLoadingConfigFile () {
+    return this._state === 'loadingConfig'
+  }
+
+  get isInError () {
+    return this._state === 'errored'
+  }
+
+  get configFilePath () {
+    assert(this._configFilePath)
+
+    return this._configFilePath
+  }
+
+  set configFilePath (configFilePath) {
+    this._configFilePath = configFilePath
+  }
+
+  setTestingType (testingType: TestingType) {
+    this._testingType = testingType
+  }
+
+  private get envFilePath () {
+    return path.join(this.options.projectRoot, 'cypress.env.json')
+  }
+
+  private get loadedConfigFile (): Partial<Cypress.ConfigOptions> | null {
+    return this._cachedLoadConfig?.initialConfig ?? null
+  }
+
+  async initializeConfig (): Promise<LoadConfigReply['initialConfig']> {
+    try {
+      this._state = 'loadingConfig'
+
+      // Clean things up for a new load
+      this.closeWatchers()
+      this._cachedLoadConfig = undefined
+      this._cachedFullConfig = undefined
+
+      const result = await this.loadConfig()
+
+      // This is necessary as there is a weird timing issue where an error occurs and the config results get loaded
+      // TODO: see if this can be !== 'errored'
+      if (this._state === 'loadingConfig') {
+        debug(`config is loaded for file`, this.configFilePath, this._testingType)
+        this.validateConfigFile(this.configFilePath, result.initialConfig)
+
+        this._state = 'loadedConfig'
+        this._cachedLoadConfig = result
+
+        this.options.onInitialConfigLoaded(result.initialConfig)
+      }
+
+      return result.initialConfig
+    } catch (error) {
+      debug(`catch %o`, error)
+      if (this._eventsIpc) {
+        this._eventsIpc.cleanupIpc()
+      }
+
+      this._state = 'errored'
+      this.closeWatchers()
+
+      throw error
+    } finally {
+      this.options.ctx.emitter.toLaunchpad()
+    }
+  }
+
+  loadTestingType () {
+    // If we have set a testingType, and it's not the "target" of the
+    // registeredEvents (switching testing mode), we need to get a fresh
+    // config IPC & re-execute the setupTestingType
+    if (this._registeredEventsTarget && this._testingType !== this._registeredEventsTarget) {
+      this.reloadConfig().catch(this.onLoadError)
+    } else if (this._eventsIpc && !this._registeredEventsTarget && this._cachedLoadConfig) {
+      this.setupNodeEvents(this._cachedLoadConfig).catch(this.onLoadError)
+    }
+  }
```

Comment:
This function is async and would be useful to await.

When this is called, it usually is a good idea to know when it's ended.

```suggestion
  async loadTestingType () {
    // If we have set a testingType, and it's not the "target" of the
    // registeredEvents (switching testing mode), we need to get a fresh
    // config IPC & re-execute the setupTestingType
    if (this._registeredEventsTarget && this._testingType !== this._registeredEventsTarget) {
      await this.reloadConfig().catch(this.onLoadError)
    } else if (this._eventsIpc && !this._registeredEventsTarget && this._cachedLoadConfig) {
      await this.setupNodeEvents(this._cachedLoadConfig).catch(this.onLoadError)
    }
  }
```